### PR TITLE
Deploy subcommand for CLI tool

### PR DESCRIPTION
**Pull request by ogoding**
Implements the `deploy` subcommand.

`deploy` cleans the relevant `target` directory, runs tests and builds the project. It then creates a `deploy` directory, zips all files (including the directory itself) in `resources`, and copies all dynamically linked libraries and binaries to `deploy`.

### CODE REVIEW

**Comment by ogoding** [comment]
I was also planning on including the option of specifying the deploy directory, but I think that could wait for some time in the future.

**Comment by White-Oak** [comment]
@ogoding #24 should allow to pass strings to subcommands, but will aso break your code, so you'll need to merge from master, if my PR is going to land (I hope it will).

**Comment by ogoding** [comment]
@White-Oak Ah, so now Vec<&str> are ArgMatches interchangeable for the commands? Cool. I'll have a look at merging from master when it lands :smile: I'll make deploy always just supply `vec!["--release", "--color=always"]` to test and build.

**Comment by White-Oak** [comment]
@ogoding those commands already supply `color=always` and instead of `--release`, you'll need to supply `release` to them (that's what they're looking for).

**Comment by ebkalderon** [comment]
@White-Oak Pull request #24 looks pretty good to me. @ogoding, what are your thoughts? Whose PR should I merge first to minimize breakage?

**Comment by White-Oak** [comment]
@ebkalderon #24 is going to be merged first, since @ogoding 's work is dependent on this.

**Comment by ebkalderon** [comment]
@ogoding All right, #24 has been merged. Go ahead and rebase! If you need any assistance, I'm sure we'd all be happy to help.

**Comment by ogoding** [comment]
@ebkalderon Merging it now, just trying to work out how to call/import the new methods from the deploy Cmd struct at the moment.

**Comment by ogoding** [comment]
@ebkalderon Should be merged without issue now! Will try to finish off the conversion from `std;:io::Error` to `&str` and then it'll all be finally done! :smile:

**Comment by LucioFranco** [comment]
@ogoding I have some work done with that. Would you like me to give you a small PR that removes the `tryio!` macro so we can just use `try!` everywhere and not have that borrow issue.

File: src/cli/subcmds/deploy.rs [hunk 1]
```diff
@@ -118,21 +118,25 @@ fn zip_dir(dir: &str, target_file: &str) -> Result<(), Error> {
...
-            tryio!(copy_binaries(&Path::new(BUILD_DIR).to_str().unwrap(), &Path::new(DEPLOY_DIR).to_str().unwrap()));
-
-            Ok(a)
-        },
-        Err(e) => Err(e),
+pub struct Cmd;
+impl AmethystCmd for Cmd {
+    /// Compresses and deploys the project as a distributable program.
+    fn execute<I: AmethystArgs>(matches: &I) -> cargo::CmdResult {
+        let cargo_args = vec!["--release"];
```
Comment (White-Oak) [inline src/cli/subcmds/deploy.rs]: Once again, build and test subcommands look for `release` argument, not `--release`.

**Comment by ogoding** [comment]
@LucioFranco That'd be great, thanks. Hopefully you don't have any conflicts, but I suspect you might.

File: src/cli/subcmds/deploy.rs (hunk 1 above)
Comment (ogoding) [inline src/cli/subcmds/deploy.rs]: Oh whoops, sorry @White-Oak. I'll fix that now.

**Comment by LucioFranco** [comment]
@ogoding I am almost ready with the new error system. Expect a PR within the next 20. (Hopefully) :)

**Comment by ogoding** [comment]
@ebkalderon PR is finally all done! Let me know if there is anything else that I should improve.

**Comment by ebkalderon** [comment]
Wonderful job! I'll try it out on my laptop tomorrow.

**Comment by ebkalderon** [comment]
@ogoding The generated `resources.zip` file structure seems incorrect when deploying an empty project on my laptop.

Currently it results in:
- `resources.zip`
  - `resources/` (folder)
    - `entities` (zero-length file)
    - `prefabs` (zero-length file)
    - `config.yml` (file)
    - `input.yml` (file)
  - `resources` (zero-length file)

It should be:
- `resources.zip`
  - `entities/` (folder)
  - `prefabs/` (folder)
  - `config.yml` (file)
  - `input.yml` (file)

**Comment by ogoding** [comment]
@ebkalderon hmm, I'll have a look at it. Probably that resources file is due to `walkdir` using the given path as the first file/folder it iterates.

**Comment by White-Oak** [comment]
AFAIK, zip can't have folders without files.

**Comment by ogoding** [comment]
@ebkalderon Just pushed a fix for those compression issues. Turns out that rust doesn't include a / character at the end of a directory's path. Also `WalkDir` starts in the current directory so I had to change it back to `fs::read_dir(path)` and use `WalkDir` for any sub directories in `resources/`.

**Comment by ebkalderon** [comment]
@ogoding Testing it now...

**Comment by ebkalderon** [comment]
@ogoding It works perfectly! Thank you so much for your effort and patience. I think this pull request is _finally_ ready to be merged.

**Pull request merged**